Add LibGuidesAZ backend for Databases A-Z list

maccabeelevine opened the pull request:
Purpose:
- Display results from a search of the LibGuides Databases A-Z list, which is a separate data source from the standard LibGuides search (of research guides).
- In particular, there should be the ability to include these results within a Combined search page.  Currently that requires the AZ list be a separate backend, since there is no way otherwise to distinguish it from the standard LibGuides data in combined.ini.

Approach:
- Like the existing LibGuides integration, we can't use the LibGuides actual API because it doesn't support search.  The [API doc itself](https://ask.springshare.com/libguides/faq/873) makes clear that the "widget" API should be used instead.
- In most cases just extend the existing LibGuides classes.

maccabeelevine committed:
Add LibGuidesAZ backend for Databases A-Z list

- In most cases just extend the existing LibGuides classes.

maccabeelevine commented:
@demiankatz I haven't yet written unit tests for this, wanted your eyes on the main code first.  Also I don't think the same "extend the base LibGuides class" approach will work for test files, or at least I've never done that myself, sit would likely just be a copy/paste of those four files.

maccabeelevine committed:
Add missing abstract method

demiankatz commented on module/VuFind/src/VuFind/Controller/LibGuidesAZController.php:
```diff
@@ -0,0 +1,70 @@
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2010.
```
Might want to update dates to reflect when the files were created.

maccabeelevine replied:
Oops.  Done.

demiankatz commented on module/VuFind/src/VuFind/Controller/LibGuidesAZController.php:
```diff
@@ -0,0 +1,70 @@
+    protected function resultScrollerActive()
+    {
+        $config = $this->serviceLocator->get(\VuFind\Config\PluginManager::class)
+            ->get('LibGuidesAZ');
```
You should probably include an example LibGuidesAZ.ini file (did you perhaps forget to commit this?).

demiankatz replied:
Note also that you could theoretically use our built in [Parent_Config] support to make LibGuidesAZ.ini inherit all or most settings from LibGuides.ini if you think that would improve maintainability.

maccabeelevine replied:
I didn't even create a separate LibGuidesAZ.ini, that line 66 was a mistake, and somehow it still picks up the configuration from LibGuides.ini.  Must be the fact I had LibGuidesAZ/Options extending LibGuides/Options.  Fixed.

I decided to take your [Parent_Config] suggestion.  Of the seven properties in LibGuides.ini, the first four (timeout, institution ID, version, baseUrl) probably have to be identical for LibGuidesAZ.  The other three (default_limit, limit_options, defaultSearch) theoretically could be different, i.e. perhaps defaultLimit if both LibGuides types are used in a Combined Search solution.

I wasn't sure whether to copy / paste and comment out all the properties into LibGuidesAZ.ini.  On the plus side it's easier to override then, on the minus you have to maintain them all in both places.  I decided not to copy them but can do so if it's VuFind practice.

demiankatz replied:
Usually our practice is to copy the settings but not the associated comments, instead putting in a note explaining where the detailed information can be found. This helps prevent things from getting out of sync over time while still providing helpful reminders of available settings. See [Search2.ini](https://github.com/vufind-org/vufind/blob/dev/config/vufind/Search2.ini) for an example of this.

maccabeelevine replied:
Ok, I copied the settings.  I think the [Parent_Config] section being present, plus its comment, makes clear where the details can be found.

demiankatz commented on module/VuFind/src/VuFind/Controller/LibGuidesAZController.php:
```diff
@@ -0,0 +1,70 @@
+        $config = $this->serviceLocator->get(\VuFind\Config\PluginManager::class)
+            ->get('LibGuidesAZ');
+        return isset($config->Record->next_prev_navigation)
+            && $config->Record->next_prev_navigation;
```
This might be cleaner as `return $config->Record->next_prev_navigation ?? false;`

(And I realize you copied this from the existing code... and I also realize that this probably serves no real purpose at all since I don't think there's a record view where this would even apply... but I just can't resist advocating for null coalescing at every opportunity ;-) ).

maccabeelevine replied:
Yeah this was a copy/paste and yeah this will probably never apply, but I agree with the principle.  Fixed.

demiankatz commented on themes/bootstrap3/templates/libguidesaz/home.phtml:
```diff
@@ -0,0 +1,2 @@
+<?php
+    include $_SERVER['CONTEXT_DOCUMENT_ROOT'] . '/../themes/bootstrap3/templates/libguides/home.phtml';
```
Is there a reason you can't `$this->render('libguides/home.phtml')` here for cleaner code?

maccabeelevine replied:
The only reason was "laminas newbie".  Fixed.

demiankatz commented on themes/bootstrap3/templates/libguidesaz/results.phtml:
```diff
@@ -0,0 +1,2 @@
+<?php
+    include $_SERVER['CONTEXT_DOCUMENT_ROOT'] . '/../themes/bootstrap3/templates/libguides/results.phtml';
```
I'd also use render here if possible. See, for example, how we do this in [search2](https://github.com/vufind-org/vufind/blob/dev/themes/bootstrap3/templates/search2/results.phtml).

maccabeelevine replied:
Done.

demiankatz commented on module/VuFind/src/VuFind/Search/Factory/LibGuidesAZBackendFactory.php:
```diff
@@ -0,0 +1,112 @@
+
+namespace VuFind\Search\Factory;
+
+use VuFindSearch\Backend\LibGuidesAZ\Backend;
```
If you want to reduce the number of spare classes you're creating, I suspect that you can eliminate all of the LibGuidesAZ classes, and just reuse equivalent LibGuides classes. Having separate classes is important inside the VuFind module since it relates to routing and parameter processing, but once you get to the search backend code, there's no consequence to reusing the same backend code multiple times (we already do this for Solr, etc.).

It looks to me like the only code that's unique to the Backend\LibGuidesAZ classes is the widget_type used in the query builder... and you could address that by adding a `defaultWidgetType` property to the class, and a `setDefaultWidgetType` method, and then have the factory set up the query builder appropriately.

There's nothing wrong with your current approach if you think there might be a reason to more explicitly differentiate functionality in the future, but if you want to reduce the number of files in the PR, I think this could be an easy win.

maccabeelevine replied:
I took this advice.  I'm not sure about whether more differentiation will be necessary in the future, but I think better to cross that bridge later than add five trivial files at this point.

demiankatz reviewed (changes requested):
Thanks again, @maccabeelevine! This looks like a great start. I just have a few comments (most significantly, about possibly reducing the number of files added to the VuFindSearch module). I haven't done a really close review of your factory refactoring yet, since I imagine it may change again in response to my comments. I'll look there more carefully once you've had a chance to respond to this current round of review.

Also note that I unfortunately can't test this myself hands-on since I don't have a LibGuides account. It might be worth reaching out to the community to see if anyone else is interested in getting involved in this review process.

demiankatz committed:
Merge branch 'dev' into libguides-az

demiankatz commented:
@maccabeelevine, I've merged the latest dev branch into this PR in an effort to get our green checkmark back (since we recently added PHP 8.2 testing requirements). I think this ball is still in your court, so please let me know when it's convenient to discuss the above review!

maccabeelevine commented:
Hi @demiankatz yes this is on my plate, hopefully next week!  Thanks for the ping.

maccabeelevine committed:
Use null coalescing

maccabeelevine committed:
Replace include statements with render()

maccabeelevine committed:
Remove VuFindSearch/Backend/LibGuidesAZ classes

Reuse the ones in LibGuides

maccabeelevine committed:
Fix copyright date for new files

maccabeelevine committed:
Add LibGuidesAZ.ini

maccabeelevine committed:
Refactor LibGuides backend factories

Now that the actual backend classes are the same, there's only a one line difference between the two factories.  More consistent then to just have LibGuidesAZ extend LibGuides, similar to all the other AZ classes, rather than have an abstract class.

maccabeelevine committed:
Fix parameter type

maccabeelevine committed:
Merge branch 'dev' into libguides-az

maccabeelevine commented:
@demiankatz Ok I addressed all the issues above.  The PHP 8 build is broken again, looks like an issue with testMergedParamBag, but I'm not clear how to resolve it and I don't write to write any AZ-specific tests until the existing ones work again.

demiankatz reviewed (changes requested):
@maccabeelevine, thanks for all the progress on this. I'm pretty sure that all the failing tests are related to the fact that you've added a new parameter to the ParamBag used in the backend, so test expectations need to be adjusted to reflect the fact that there's an extra value now. Let me know if you need more detailed support sorting it out!

demiankatz commented:
To be a little more precise, @maccabeelevine, I think the issue is that all the expectations like this one:

https://github.com/vufind-org/vufind/blob/dev/module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/LibGuides/BackendTest.php#L169

Need to have the widget_type value added to them.

maccabeelevine committed:
Copy settings from LibGuides to LibGuidesAZ

Commented out, since Parent_Config is present

maccabeelevine committed:
Tweak comment

demiankatz commented on config/vufind/LibGuidesAZ.ini:
```diff
@@ -0,0 +1,16 @@
+[Parent_Config]
+relative_path = ./LibGuides.ini
+
+; ---------- config.ini settings ----------
```
I don't think you need this comment; these settings don't come from config.ini. I'd either change this to say "LibGuides.ini" or remove it entirely. Given the Parent_Config settings right above, removing it is probably fine.

maccabeelevine replied:
Yeah typo, I meant Libguides.ini obviously.  But you're right it's redundant, I removed it.

demiankatz replied:
Thanks!

demiankatz commented on config/vufind/LibGuidesAZ.ini:
```diff
@@ -0,0 +1,16 @@
+;baseUrl = https://lgapi.libapps.com/widgets.php
+;default_limit        = 20
+;limit_options        = 10,20,40,60,80,100
+;defaultSearch = "general"
```
You might want to add an end-of-file line break here for consistency with most other files in the project.

maccabeelevine replied:
Done!

demiankatz reviewed (changes requested):
Thanks for the .ini update -- a couple of very minor nitpicks for you. :-)

I'll give this some hands-on testing once the test suite is passing. Let me know if I can help with anything else in the meantime!

maccabeelevine committed:
Remove comment mistake

maccabeelevine committed:
Fix BackendTest

Needed 'widget_type' param

demiankatz commented on module/VuFind/src/VuFind/Search/LibGuides/Options.php:
```diff
@@ -40,14 +40,16 @@
  */
 class Options extends \VuFind\Search\Base\Options
 {
+    protected string $iniName = 'LibGuides';
```
This should get a comment on it, like:

```
/**
 * Name of .ini file to use for search and facet settings.
 *
 * @var string
 */
```

maccabeelevine replied:
php-cs-fixer and phpcs-console didn't catch this one!  Fixed.

demiankatz replied:
Yes, for some reason, the style checkers don't complain about this -- it just creates a warning in the PHPDocumentor output. In any case, thanks for the fix!

demiankatz commented on module/VuFind/src/VuFind/Search/LibGuidesAZ/Options.php:
```diff
@@ -0,0 +1,56 @@
+ */
+class Options extends \VuFind\Search\LibGuides\Options
+{
+    protected string $iniName = 'LibGuidesAZ';
```
This needs the same comment as the other Options class.

maccabeelevine replied:
Fixed!

demiankatz reviewed (changes requested):
I did one last review of this before testing, and I noticed one trivial thing in need of fixing: a missing doc comment on a couple of class properties.

I'll go ahead with my testing now since fixing this problem has no functional effect on the code. I'll report on my findings shortly! :-)

maccabeelevine committed:
Fix QueryBuilder construction in refactoring

maccabeelevine committed:
Add missing comment to $iniName

maccabeelevine committed:
Fix linting

demiankatz reviewed (approved):
Looks good, @maccabeelevine! Merging now.

dltj commented:
@maccabeelevine : Do you have an example of how this looks in the Lehigh VuFind interface? I looked for [ABI/INFORM](https://asa.lib.lehigh.edu/Search/Results?lookfor=ABI%2FINFORM&type=AllFields&submit=Search) and [Access Science](https://asa.lib.lehigh.edu/Search/Results?lookfor=Access+Science&type=AllFields&limit=20), for instance, and those look like FOLIO records.

maccabeelevine commented:
> @maccabeelevine : Do you have an example of how this looks in the Lehigh VuFind interface? I looked for [ABI/INFORM](https://asa.lib.lehigh.edu/Search/Results?lookfor=ABI%2FINFORM&type=AllFields&submit=Search) and [Access Science](https://asa.lib.lehigh.edu/Search/Results?lookfor=Access+Science&type=AllFields&limit=20), for instance, and those look like FOLIO records.

@dltj Sorry no it's not in production yet at Lehigh.  Happy to screenshare if you like.
